### ISSUE TITLE
BeanCreationException upgrading from 2.1.32 to 2.1.33 or later

### ISSUE THREAD

**Issue by wfhartford** [adobe/S3Mock#275]
I upgraded s3mock to the latest version and now get an exception when starting. The breakage seems to be from version 2.1.32 to 2.1.33 with all later versions exhibiting the same behaviour.

S3Mock is being started in a test class using the following code (kotlin):
```kotlin
S3MockApplication.start(mapOf(
        S3MockApplication.PROP_HTTPS_PORT to S3MockApplication.RANDOM_PORT,
        S3MockApplication.PROP_HTTP_PORT to S3MockApplication.RANDOM_PORT,
        S3MockApplication.PROP_INITIAL_BUCKETS to "bucket"
    ))
```

This results in the following exception:
```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'connectionFactory' defined in class path resource [org/springframework/boot/autoconfigure/r2dbc/ConnectionFactoryConfigurations$Generic.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [io.r2dbc.spi.ConnectionFactory]: Factory method 'connectionFactory' threw exception; nested exception is org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryBuilder$ConnectionFactoryBeanCreationException: Failed to determine a suitable R2DBC Connection URL
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:657)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:637)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1341)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1181)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:556)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:516)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:324)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:322)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:897)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:879)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:551)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:143)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:755)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:402)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:312)
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:140)
	at com.adobe.testing.s3mock.S3MockApplication.start(S3MockApplication.java:214)
        [redacted]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [io.r2dbc.spi.ConnectionFactory]: Factory method 'connectionFactory' threw exception; nested exception is org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryBuilder$ConnectionFactoryBeanCreationException: Failed to determine a suitable R2DBC Connection URL
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:652)
	... 50 more
Caused by: org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryBuilder$ConnectionFactoryBeanCreationException: Failed to determine a suitable R2DBC Connection URL
	at org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryBuilder$ConnectionFactoryOptionsInitializer.connectionFactoryBeanCreationException(ConnectionFactoryBuilder.java:227)
	at org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryBuilder$ConnectionFactoryOptionsInitializer.initializeOptions(ConnectionFactoryBuilder.java:156)
	at org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryBuilder.of(ConnectionFactoryBuilder.java:62)
	at org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryConfigurations.createConnectionFactory(ConnectionFactoryConfigurations.java:51)
	at org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryConfigurations$Generic.connectionFactory(ConnectionFactoryConfigurations.java:90)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	... 51 more
```

Looking through the release notes and code diff between the two versions did not reveal anything that I would expect could cause an issue like this.

**Comment by afranken** [comment]
Not reproducible.
See #276 which adds a unit test that does what your code does (only it's in Java).

2.1.32 to 2.1.33 was actually a big change in dependencies. I updated almost every dependency we have, most notably the Spring Boot dependency, which in turn updates the Spring Framework.

Maybe I should have changed the minor or major version instead of the patch version for that release - if you're using the S3MockApplication directly, or one of the testsupport modules in your code from within a Spring application, this may break your project. :(

My guess: your project has some Spring Framework / Spring Boot dependency that includes an auto configuration for R2DBC. The newer Spring Boot/Framework version we provide in S3Mock now detects the autoconfiguration and tries to instantiate the beans as part of the S3Mock.

**Comment by afranken** [comment]
Can you run Docker in your build?
I just added a testsupport module for TestContainers (will be released later today).
If you start through the Java API, there will always be the risk that some dependency in your project will break the S3Mock, and vice versa.

Using Docker in your tests will insulate both applications dependencies and application life cycles.

**Comment by wfhartford** [comment]
Ahhh, I was wondering how much was hiding behind those parent pom changes. Thanks for looking into this. I noticed your testcontainer support when I was investigating and thought that would be a much better solution. I'll go down that path instead.

The project I'm working in doesn't include any spring dependencies other than the ones pulled in by s3mock. The only dependencies which are changed from what s3mock would have by default are log4j (upgraded to 2.14.1) and jackson (upgraded to 2.12.4). My project does use jooq, which depends on `io.r2dbc:r2dbc-spi:0.9.0.M1`, I wonder if that might have caused the issue.

If you want to spend any more time looking into this issue, I'm happy to help where I can. Otherwise, this can probably be closed. Either way, I'll probably move to the testcontainers option.

**Comment by afranken** [comment]
> My project does use jooq, which depends on io.r2dbc:r2dbc-spi:0.9.0.M1, I wonder if that might have caused the issue.

Spring is always very clever on which parts of its own configuration to use and which beans to instantiate.
There's a  `@ConditionalOnClass(ConnectionFactory.class)` (on `io.r2dbc.spi.ConnectionFactory` to be exact) that was triggered because the class is available in your classpath when starting the S3MockApplication:
https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/r2dbc/R2dbcAutoConfiguration.java#L38

I just released 2.2.0 for TestContainers support. :)
